raft/node.go

```diff
@@ -503,6 +503,17 @@ func (n *node) stepWithWaitOption(ctx context.Context, m pb.Message, wait bool)
 func (n *node) Ready() <-chan Ready { return n.readyc }
 
 func (n *node) Advance() {
+	ents := n.rd.Entries
+
+	// Provide a feedback that the `Entries` has already been successfully
+	// saved by the user. Note that only leader needs to do this, because
+	// a follower will not send the feedback to leader until it finishes
+	// saving the `Entries`.
+	if n.isLeader() && (len(ents) > 1 || (len(ents) == 1 && len(ents[0].Data) > 0)) {
```

Comment:
Could you, please describe why len(ents) >= 1  would be not sufficient here,
it is, what case and why we want to filter out here.

Reply:
Good point, I was planning to add more comment as well. Will update soon. thx

Reply:
// An empty entry is generated in the following two cases:
	//    1. A new leader is elected.
	//    2. enter into joint configuration.
	// We need to ignore above two cases because they are already committed. Note
	// that it's safe to commit the empty entry immediately after appending the entry
	// because it can only happen for multi-member cluster.

Reply:
Updated.

@ptabor FYI. https://github.com/ahrtr/etcd/blob/a79941679f401186dc29e10d4e86a53590c3d9d6/raft/raft.go#L643-L645

Reply:
Just updated the comment to avoid confusion,

```
	// An empty entry is generated in the following two cases:
	//    1. a new leader is elected.
	//    2. enter into joint configuration.
	// We need to ignore above two cases because the `pr.Match` is updated
	// immediately after appending the empty entry. Note that it's safe to do
	// so because it can only happen for multi-member cluster. Please also refer to
	// https://github.com/ahrtr/etcd/blob/a79941679f401186dc29e10d4e86a53590c3d9d6/raft/raft.go#L643-L645
```